others/freivalds_algorithm.cpp

```diff
@@ -0,0 +1,169 @@
+/**
+ * @file
+ * @brief Implementation of [Freivalds' algorithm]
+ * (https://en.wikipedia.org/wiki/Freivalds%27_algorithm)
+ * @author [Darshana Sarma](https://github.com/Darshana-Sarma)
+ * @details
+ * Given three n × n matrices A, B, and
+ * C, a general problem is to verify whether AB=C
+ *
+ * The time Complexity of the algorithm is O(n^{2}).
+ */
+#include <array>     //   for std::array
+#include <cassert>   //   for assert
+#include <iostream>  //   for io operations
+
+/**
+ * @namespace others
+ * @brief Other algorithms
+ */
+namespace others {
+/**
+ * @namespace freivalds
+ * @brief Functions for [Freivalds']
+ * (https://en.wikipedia.org/wiki/Freivalds%27_algorithm) algorithm
+ */
+namespace freivalds {
+/**
+ * The function will check AB=C by applying Freidvald's algorithm 
+ * @tparam N dimension of square matrix NxN
+ * @param A square matrix A
+ * @param B square matrix B
+ * @param C resultant square matrix C
+ * @returns boolean value
+ */
+template <std::size_t N>
+bool CoppersmithAlgo(std::array<std::array<double, N>, N> A,
+                     std::array<std::array<double, N>, N> B,
+                     std::array<std::array<double, N>, N> C) {
+    // Printing matrix A
+    std::cout << "\nThe first matrix is:\n ";
+    for (int i = 0; i < N; i++) {
+        for (int j = 0; j < N; j++) {
+            std::cout << A[i][j] << "\t";
+        }
+        std::cout << "\n";
+    }
+
+    // Printing matrix B
+    std::cout << "\nThe Second matrix is:\n ";
+    for (int i = 0; i < N; i++) {
+        for (int j = 0; j < N; j++) {
+            std::cout << B[i][j] << "\t";
+        }
+        std::cout << "\n";
+    }
+
+    // Printing matrix C
+    std::cout << "The resultant matrix: \n";
+    for (int i = 0; i < N; i++) {
+        for (int j = 0; j < N; j++) {
+            std::cout << C[i][j] << "\t";
+        }
+        std::cout << "\n";
+    }
+    // generate a random nx1 vector r containing 0/1 elements

```

Comment:
If you use the algorithm with a random vector of length `n`, you get a time complexity `O(n^3)` which is as slow as multiplying the matrices altogether (while being a bit worse because it introduces a chance of failure).

I believe the real strength of the algorithm is using it with shorter (constant-sized) random vectors of length k that according to the linked wikipedia article lead to less than `1 / 2^k` probability of failure.